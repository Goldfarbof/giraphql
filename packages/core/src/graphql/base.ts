import { GraphQLType } from 'graphql';
import BasePlugin from '../plugin';
import BuildCache from '../build-cache';

export default abstract class BaseType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string | keyof Types['Input'] | keyof Types['Output'],
  Shape,
  InputShape = Shape,
  MatchShape = Shape
> {
  typename: Name;

  shape?: Shape;

  inputShape?: InputShape;

  matchShape?: MatchShape;

  abstract kind: 'Object' | 'Union' | 'Interface' | 'Enum' | 'Scalar' | 'InputObject';

  constructor(name: Name, shape?: Shape) {
    this.typename = name;
    this.shape = shape;
  }

  abstract buildType(cache: BuildCache<Types>, plugins: BasePlugin<Types>[]): GraphQLType;
}
