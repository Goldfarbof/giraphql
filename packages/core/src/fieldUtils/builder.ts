import { TypeParam, CompatibleTypes, FieldOptionsFromKind, FieldNullability } from '../types';
import Field from '../graphql/field';
import RootFieldBuilder from './root';

export default class FieldBuilder<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  ParentShape,
  Kind extends 'Object' | 'Interface' = 'Object' | 'Interface'
> extends RootFieldBuilder<Types, ParentShape, Kind> {
  exposeBoolean<
    Name extends CompatibleTypes<Types, ParentShape, 'Boolean', Nullable>,
    Nullable extends FieldNullability<'Boolean'> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, 'Boolean', Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, 'Boolean'> {
    return this.exposeField<'Boolean', Nullable, Name>(name, { ...options, type: 'Boolean' });
  }

  exposeFloat<
    Name extends CompatibleTypes<Types, ParentShape, 'Float', Nullable>,
    Nullable extends FieldNullability<'Float'> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, 'Float', Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, 'Float'> {
    return this.exposeField<'Float', Nullable, Name>(name, { ...options, type: 'Float' });
  }

  exposeID<
    Name extends CompatibleTypes<Types, ParentShape, 'ID', Nullable>,
    Nullable extends FieldNullability<'ID'> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, 'ID', Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, 'ID'> {
    return this.exposeField<'ID', Nullable, Name>(name, { ...options, type: 'ID' });
  }

  exposeInt<
    Name extends CompatibleTypes<Types, ParentShape, 'Int', Nullable>,
    Nullable extends FieldNullability<'Int'> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, 'Int', Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, 'Int'> {
    return this.exposeField<'Int', Nullable, Name>(name, { ...options, type: 'Int' });
  }

  exposeString<
    Name extends CompatibleTypes<Types, ParentShape, 'String', Nullable>,
    Nullable extends FieldNullability<'String'> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, 'String', Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, 'String'> {
    return this.exposeField<'String', Nullable, Name>(name, { ...options, type: 'String' });
  }

  exposeBooleanList<
    Name extends CompatibleTypes<Types, ParentShape, ['Boolean'], Nullable>,
    Nullable extends FieldNullability<['Boolean']> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, ['Boolean'], Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, ['Boolean']> {
    return this.exposeField<['Boolean'], Nullable, Name>(name, { ...options, type: ['Boolean'] });
  }

  exposeFloatList<
    Name extends CompatibleTypes<Types, ParentShape, ['Float'], Nullable>,
    Nullable extends FieldNullability<['Float']> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, ['Float'], Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, ['Float']> {
    return this.exposeField<['Float'], Nullable, Name>(name, { ...options, type: ['Float'] });
  }

  exposeIDList<
    Name extends CompatibleTypes<Types, ParentShape, ['ID'], Nullable>,
    Nullable extends FieldNullability<['ID']> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, ['ID'], Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, ['ID']> {
    return this.exposeField<['ID'], Nullable, Name>(name, { ...options, type: ['ID'] });
  }

  exposeIntList<
    Name extends CompatibleTypes<Types, ParentShape, ['Int'], Nullable>,
    Nullable extends FieldNullability<['Int']> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, ['Int'], Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, ['Int']> {
    return this.exposeField<['Int'], Nullable, Name>(name, { ...options, type: ['Int'] });
  }

  exposeStringList<
    Name extends CompatibleTypes<Types, ParentShape, ['String'], Nullable>,
    Nullable extends FieldNullability<['String']> = false
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, ['String'], Nullable, {}, Kind, ParentShape>,
      'resolve' | 'type'
    >,
  ): Field<{}, Types, ['String']> {
    return this.exposeField<['String'], Nullable, Name>(name, { ...options, type: ['String'] });
  }

  expose<
    Type extends TypeParam<Types>,
    Nullable extends boolean,
    Name extends CompatibleTypes<Types, ParentShape, Type, Nullable>
  >(
    name: Name,
    options: Omit<
      FieldOptionsFromKind<Types, ParentShape, Type, Nullable, {}, Kind, ParentShape>,
      'resolve'
    >,
  ): Field<{}, Types, Type> {
    return this.exposeField(name, options);
  }
}
