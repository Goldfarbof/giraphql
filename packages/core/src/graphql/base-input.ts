import BaseType from './base';

export default abstract class BaseInputType<Shape = unknown, InputShape = Shape> extends BaseType<
  Shape
> {
  shape?: Shape;

  inputShape?: InputShape;

  abstract kind: 'Enum' | 'Scalar' | 'InputObject';
}
