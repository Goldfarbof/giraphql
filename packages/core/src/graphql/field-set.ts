import { FieldsShape } from '../types';

export default class FieldSet<Types extends GiraphQLSchemaTypes.TypeInfo, Shape = unknown> {
  kind: 'FieldSet' = 'FieldSet';

  forType: string;

  shape: FieldsShape<any, any>;

  constructor(type: string, shape: FieldsShape<Types, Shape>) {
    this.forType = type;
    this.shape = shape;
  }
}
