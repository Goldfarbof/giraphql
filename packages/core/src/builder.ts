/* eslint-disable no-dupe-class-members */
import { GraphQLSchema, GraphQLScalarType, GraphQLDirective } from 'graphql';
import {
  ImplementedType,
  EnumValues,
  InputFields,
  InputShapeFromFields,
  ObjectName,
  InterfaceName,
  ScalarName,
  ResolverMap,
  CompatibleInterfaceParam,
  FieldMap,
  ObjectFieldsShape,
  QueryFieldsShape,
  MutationFieldsShape,
  SubscriptionFieldsShape,
  InterfaceFieldsShape,
  ObjectFieldThunk,
  InterfaceFieldThunk,
  QueryFieldThunk,
  MutationFieldThunk,
  SubscriptionFieldThunk,
} from './types';
import ObjectType from './graphql/object';
import UnionType from './graphql/union';
import InputObjectType from './graphql/input';
import InterfaceType from './graphql/interface';
import EnumType from './graphql/enum';
import ScalarType from './graphql/scalar';
import BuildCache from './build-cache';
import {
  InputFieldBuilder,
  ObjectFieldBuilder,
  QueryFieldBuilder,
  MutationFieldBuilder,
  SubscriptionFieldBuilder,
  InterfaceFieldBuilder,
} from '.';
import { BasePlugin, mergePlugins } from './plugins';
import QueryType from './graphql/query';
import MutationType from './graphql/mutation';
import SubscriptionType from './graphql/subscription';

export default class SchemaBuilder<Types extends GiraphQLSchemaTypes.TypeInfo> {
  private plugin: Required<BasePlugin>;

  private types = new Map<string, ImplementedType>();

  private pendingFields = new Map<string, FieldMap[]>();

  private fields = new Map<string, FieldMap[]>();

  constructor(options: { plugins?: BasePlugin[] } = {}) {
    this.plugin = mergePlugins(options.plugins ?? []);
  }

  objectType<
    Interfaces extends CompatibleInterfaceParam<Types, Types['Object'][Type]>[],
    Type extends ObjectName<Types>
  >(
    name: Type,
    options:
      | GiraphQLSchemaTypes.ObjectTypeOptions<Types, Types['Object'][Type]>
      | GiraphQLSchemaTypes.ObjectTypeWithInterfaceOptions<
          Types,
          Types['Object'][Type],
          Interfaces
        >,
    shape?: ObjectFieldsShape<Types, Types['Object'][Type]>,
  ) {
    const type = this.addType(
      new ObjectType<Types>(
        name,
        options as GiraphQLSchemaTypes.ObjectTypeOptions<Types, Types['Object'][Type]>,
      ),
    );

    if (shape) {
      this.addFields(name, shape(new ObjectFieldBuilder(name)));
    }

    if (options.shape) {
      this.addFields(name, options.shape(new ObjectFieldBuilder(name)));
    }

    return type;
  }

  objectFields<Type extends ObjectName<Types>>(
    name: Type,
    shape: ObjectFieldsShape<Types, Types['Object'][Type]>,
  ) {
    return this.addFields(name, shape(new ObjectFieldBuilder(name)));
  }

  objectField<Type extends ObjectName<Types>>(
    name: Type,
    fieldName: string,
    field: ObjectFieldThunk<Types, Types['Object'][Type]>,
  ) {
    return this.addFields(name, { [fieldName]: field(new ObjectFieldBuilder(name)) });
  }

  queryType(options: GiraphQLSchemaTypes.QueryTypeOptions<Types>, shape?: QueryFieldsShape<Types>) {
    const type = this.addType(new QueryType<Types>(options));

    if (shape) {
      this.addFields('Query', shape(new QueryFieldBuilder()));
    }

    if (options.shape) {
      this.addFields('Query', options.shape(new QueryFieldBuilder()));
    }

    return type;
  }

  queryFields(shape: QueryFieldsShape<Types>) {
    return this.addFields('Query', shape(new QueryFieldBuilder()));
  }

  queryField(name: string, field: QueryFieldThunk<Types>) {
    return this.addFields('Query', { [name]: field(new QueryFieldBuilder()) });
  }

  mutationType(
    options: GiraphQLSchemaTypes.MutationTypeOptions<Types>,
    shape?: MutationFieldsShape<Types>,
  ) {
    const type = this.addType(new MutationType<Types>(options));

    if (shape) {
      this.addFields('Mutation', shape(new MutationFieldBuilder()));
    }

    if (options.shape) {
      this.addFields('Mutation', options.shape(new MutationFieldBuilder()));
    }

    return type;
  }

  mutationFields(shape: MutationFieldsShape<Types>) {
    return this.addFields('Mutation', shape(new MutationFieldBuilder()));
  }

  mutationField(name: string, field: MutationFieldThunk<Types>) {
    return this.addFields('Mutation', { [name]: field(new MutationFieldBuilder()) });
  }

  subscriptionType(
    options: GiraphQLSchemaTypes.SubscriptionTypeOptions<Types>,
    shape?: SubscriptionFieldsShape<Types>,
  ) {
    const type = this.addType(new SubscriptionType<Types>(options));

    if (shape) {
      this.addFields('Subscription', shape(new SubscriptionFieldBuilder()));
    }

    if (options.shape) {
      this.addFields('Subscription', options.shape(new SubscriptionFieldBuilder()));
    }

    return type;
  }

  subscriptionFields(shape: SubscriptionFieldsShape<Types>) {
    return this.addFields('Subscription', shape(new SubscriptionFieldBuilder()));
  }

  subscriptionField(name: string, field: SubscriptionFieldThunk<Types>) {
    return this.addFields('Subscription', { [name]: field(new SubscriptionFieldBuilder()) });
  }

  args<Shape extends InputFields<Types>>(
    shape: (t: GiraphQLSchemaTypes.InputFieldBuilder<Types>) => Shape,
  ) {
    return shape(new InputFieldBuilder<Types>());
  }

  interfaceType<Type extends InterfaceName<Types>>(
    name: Type,
    options: GiraphQLSchemaTypes.InterfaceTypeOptions<Types, Types['Interface'][Type]>,
    shape?: InterfaceFieldsShape<Types, Types['Interface'][Type]>,
  ) {
    const type = this.addType(new InterfaceType<Types, Type>(name, options));

    if (shape) {
      this.addFields(name, shape(new InterfaceFieldBuilder(name)));
    }

    if (options.shape) {
      this.addFields(name, options.shape(new InterfaceFieldBuilder(name)));
    }

    return type;
  }

  interfaceFields<Type extends InterfaceName<Types>>(
    name: Type,
    shape: InterfaceFieldsShape<Types, Types['Interface'][Type]>,
  ) {
    return this.addFields(name, shape(new InterfaceFieldBuilder(name)));
  }

  interfaceField<Type extends InterfaceName<Types>>(
    name: Type,
    fieldName: string,
    field: InterfaceFieldThunk<Types, Types['Interface'][Type]>,
  ) {
    return this.addFields(name, { [fieldName]: field(new InterfaceFieldBuilder(name)) });
  }

  unionType<Member extends ObjectName<Types>, Name extends string>(
    name: Name,
    options: GiraphQLSchemaTypes.UnionOptions<Types, Member>,
  ) {
    return this.addType(new UnionType<Types, Member>(name, options));
  }

  enumType<Name extends string, Values extends EnumValues>(
    name: Name,
    options: GiraphQLSchemaTypes.EnumTypeOptions<Values>,
  ) {
    return this.addType(new EnumType(name, options));
  }

  scalarType<Name extends ScalarName<Types>>(
    name: Name,
    options: GiraphQLSchemaTypes.ScalarOptions<
      Types['Scalar'][Name]['Input'],
      Types['Scalar'][Name]['Output']
    >,
  ) {
    return this.addType(
      new ScalarType<Types['Scalar'][Name]['Input'], Types['Scalar'][Name]['Output']>(
        name,
        options,
      ),
    );
  }

  addScalarType<Name extends ScalarName<Types>>(name: Name, scalar: GraphQLScalarType) {
    return this.addType(
      new ScalarType<Types['Scalar'][Name]['Input'], Types['Scalar'][Name]['Output']>(name, {
        description: scalar.description ?? undefined,
        serialize: scalar.serialize,
        parseLiteral: scalar.parseLiteral,
        parseValue: scalar.parseValue,
        extensions: scalar.extensions ?? undefined,
      }),
    );
  }

  inputType<Name extends string, Fields extends InputFields<Types>>(
    name: Name,
    options: GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>,
  ): InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name>;

  inputType<Name extends string, Fields extends InputFields<Types>>(
    name: Name,
    options: Omit<GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>, 'shape'>,
    shape: (t: GiraphQLSchemaTypes.InputFieldBuilder<Types>) => Fields,
  ): InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name>;

  inputType<Name extends string, Fields extends InputFields<Types>>(
    name: Name,
    options: GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>,
    shape?: (t: GiraphQLSchemaTypes.InputFieldBuilder<Types>) => Fields,
  ): InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name> {
    if (shape && options.shape) {
      throw new Error(`Received multiple field shape functions for InputObjectType ${name}`);
    }

    return this.addType(
      new InputObjectType<Types, InputShapeFromFields<Types, Fields, undefined>, Name>(name, {
        ...options,
        shape: shape ?? options.shape,
      }),
    );
  }

  // Temporarily alias old names for compatibility
  createQueryType = this.queryType;

  createMutationType = this.mutationType;

  createSubscriptionType = this.subscriptionType;

  createObjectType = this.objectType;

  createInterfaceType = this.interfaceType;

  createQueryFields = this.queryFields;

  createMutationFields = this.mutationFields;

  createSubscriptionFields = this.subscriptionFields;

  createObjectFields = this.objectFields;

  createInterfaceFields = this.interfaceFields;

  createUnionType = this.unionType;

  createEnumType = this.enumType;

  createScalarType = this.addScalarType;

  createInputType = this.inputType;

  createArgs = this.args;

  toSchema({
    directives,
    extensions,
    mocks,
  }: {
    directives?: readonly GraphQLDirective[];
    extensions?: Record<string, unknown>;
    mocks?: ResolverMap;
  } = {}) {
    this.plugin.beforeBuild(this);

    if (this.pendingFields.size) {
      throw new Error(
        `Fields defined without defining type (${[...this.pendingFields.keys()].join(', ')}).`,
      );
    }

    const buildCache = new BuildCache([...this.types.values()], this.plugin, {
      fieldDefinitions: this.fields,
      mocks,
    });

    buildCache.buildAll();

    const builtTypes = [...buildCache.types.values()].map((entry) => entry.built);

    const schema = new GraphQLSchema({
      query: buildCache.has('Query')
        ? buildCache.getEntryOfType('Query', 'Query').built
        : undefined,
      mutation: buildCache.has('Mutation')
        ? buildCache.getEntryOfType('Mutation', 'Mutation').built
        : undefined,
      subscription: buildCache.has('Subscription')
        ? buildCache.getEntryOfType('Subscription', 'Subscription').built
        : undefined,
      extensions,
      directives: directives as GraphQLDirective[],
      types: builtTypes,
    });

    this.plugin.afterBuild(schema, this);

    return schema;
  }

  addType<T extends ImplementedType>(type: T) {
    this.types.set(type.typename, type);

    this.plugin.onType(type, this);

    const pending = this.pendingFields.get(type.typename);

    if (pending) {
      this.pendingFields.delete(type.typename);
      for (const fields of pending) {
        this.addFields(type.typename, fields);
      }
    }

    return type;
  }

  addFields(typename: string, fields: FieldMap) {
    const type = this.types.get(typename);

    if (type) {
      if (this.fields.has(typename)) {
        this.fields.get(typename)!.push(fields);
      } else {
        this.fields.set(typename, [fields]);
      }

      Object.keys(fields).forEach((fieldName) => {
        this.plugin.onField(type, fieldName, fields[fieldName], this);
      });
    } else if (this.pendingFields.has(typename)) {
      this.pendingFields.get(typename)!.push(fields);
    } else {
      this.pendingFields.set(typename, [fields]);
    }

    return fields;
  }
}
