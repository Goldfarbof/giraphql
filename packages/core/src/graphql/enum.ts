import { GraphQLEnumType, GraphQLEnumValueConfigMap, GraphQLEnumValueConfig } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import BaseType from './base';
import { EnumValues, ShapeFromEnumValues } from '../types';

export default class EnumType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string,
  Values extends EnumValues = EnumValues,
  Shape extends string = ShapeFromEnumValues<Values>
> extends BaseType<Types, Name, Shape> {
  values: GraphQLEnumValueConfigMap;

  description?: string;

  kind: 'Enum' = 'Enum';

  options: GiraphQLSchemaTypes.EnumTypeOptions<EnumValues>;

  constructor(name: Name, options: GiraphQLSchemaTypes.EnumTypeOptions<Values>) {
    super(name);

    this.options = options;

    this.values = Array.isArray(options.values)
      ? fromEntries(options.values.map(key => [key, {}]))
      : fromEntries(
          Object.entries(options.values)
            .map(([key, value]) => {
              if (value && typeof value === 'object') {
                return [key, value];
              }

              if (typeof value === 'string') {
                return [value, {}];
              }

              return null;
            })
            .filter(Boolean) as [string, GraphQLEnumValueConfig][],
        );

    this.description = options.description;
  }

  buildType() {
    return new GraphQLEnumType({
      description: this.description,
      name: this.typename,
      values: this.values,
    });
  }
}
