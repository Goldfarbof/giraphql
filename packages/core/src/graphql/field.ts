/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { GraphQLFieldConfig, GraphQLFieldConfigArgumentMap } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import {
  TypeParam,
  InputFields,
  ShapeFromTypeParam,
  FieldNullability,
  FieldOptionsFromKind,
} from '../types';
import { typeFromParam, buildArg } from '../utils';
import BaseType from './base';
import BasePlugin from '../plugin';
import { BuildCache } from '..';

export default class Field<
  Args extends InputFields<Types>,
  Types extends GiraphQLSchemaTypes.TypeInfo,
  ParentType extends TypeParam<Types>,
  Type extends TypeParam<Types>,
  Nullable extends FieldNullability<Types, Type> = FieldNullability<Types, Type>,
  Extends extends string | null = null,
  Kind extends 'Object' | 'Interface' | 'Root' | 'Subscription' =
    | 'Object'
    | 'Interface'
    | 'Root'
    | 'Subscription',
  Options extends FieldOptionsFromKind<
    Types,
    ParentType,
    Type,
    Nullable,
    Args,
    Kind
  > = FieldOptionsFromKind<Types, ParentType, Type, Nullable, Args, Kind>
> {
  shape?: ShapeFromTypeParam<Types, Type, true>;

  nullable: Nullable;

  args: Args = {} as Args;

  extendsField: Extends;

  type: Type;

  options: Options;

  parentTypename: string;

  constructor(
    options: Options & {
      extendsField?: Extends;
    },
    parentTypename: string,
  ) {
    this.options = options;
    this.nullable = options.nullable || (false as Nullable);
    this.args = options.args ? options.args! : ({} as Args);
    this.extendsField = options.extendsField || (null as Extends);
    this.type = options.type;
    this.parentTypename = parentTypename;
  }

  protected buildArgs(cache: BuildCache<Types>): GraphQLFieldConfigArgumentMap {
    return fromEntries(
      Object.keys(this.args).map(key => {
        const arg = this.args[key];

        return [
          key,
          {
            description:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? undefined
                : arg.description,
            required:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? false
                : arg.required || false,
            type: buildArg(arg, cache),
            defaultValue:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? undefined
                : arg.default,
          },
        ];
      }),
    );
  }

  build(
    name: string,
    cache: BuildCache<Types>,
    plugins: BasePlugin<Types>[],
  ): GraphQLFieldConfig<unknown, unknown> {
    const baseConfig: GraphQLFieldConfig<unknown, unknown> = {
      args: this.buildArgs(cache),
      description: this.options.description,
      resolve:
        cache.resolverMock(this.parentTypename, name) ||
        (this.options as { resolve?: (...args: unknown[]) => unknown }).resolve ||
        (() => {
          throw new Error(`Not implemented: No resolver found for ${this.parentTypename}.${name}`);
        }),
      subscribe:
        cache.subscribeMock(this.parentTypename, name) ||
        (this.options as { subscribe?: (...args: unknown[]) => unknown }).subscribe,
      type: typeFromParam(this.type, cache, this.nullable),
      extensions: this.options.extensions,
    };

    return plugins.reduce((config, plugin) => {
      return plugin.updateFieldConfig
        ? plugin.updateFieldConfig(name, this as any, config, cache)
        : config;
    }, baseConfig);
  }
}
