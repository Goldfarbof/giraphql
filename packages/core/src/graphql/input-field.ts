import { GraphQLInputFieldConfig, GraphQLArgumentConfig } from 'graphql';
// @ts-ignore
import { inputTypeFromParam } from '../utils';
import { BuildCache, InputType, SchemaTypes, partialInputShapeKey, PartialInput } from '..';

export default class InputField<T> implements PartialInput {
  kind: 'InputField' = 'InputField';

  [partialInputShapeKey]: T;

  required: boolean | { list: boolean; items: boolean };

  type: InputType<SchemaTypes> | [InputType<SchemaTypes>];

  options: GiraphQLSchemaTypes.InputOptions;

  constructor(options: GiraphQLSchemaTypes.InputOptions) {
    this.options = options;
    this.required = options.required ?? false;
    this.type = options.type;
  }

  // This is currently used to for fields on Input Objects, and Arguments on Fields
  build(cache: BuildCache): GraphQLInputFieldConfig & GraphQLArgumentConfig {
    return {
      description: this.options.description,
      defaultValue: this.options.defaultValue,
      extensions: this.options.extensions,
      type: inputTypeFromParam(this.type, cache, this.required),
    };
  }
}
