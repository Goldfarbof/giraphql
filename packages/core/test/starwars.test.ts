import { printSchema, execute } from 'graphql';
import gql from 'graphql-tag';
import starwarsSchema from './examples/starwars/schema';

import Character from './examples/starwars/schema/character';
import Droid from './examples/starwars/schema/droid';
import { Episode, MoreEpisodes } from './examples/starwars/schema/episode';
import Human from './examples/starwars/schema/human';
import Query, { extraQueryFields } from './examples/starwars/schema/query';

import builder from './examples/starwars/builder';

describe('starwars example', () => {
  test('generates expected schema', () => {
    expect(printSchema(starwarsSchema)).toMatchSnapshot();
  });

  describe('queries', () => {
    test('query some stuff', async () => {
      const query = gql`
        query {
          droid(id: 2001) {
            name
            primaryFunction
            friends {
              name
              appearsIn(id: 4)
              ... on Human {
                homePlanet
              }
              ... on Droid {
                primaryFunction
              }
            }
          }
          human(id: 1001) {
            name
            homePlanet
            friends {
              name
              appearsIn(id: 4)
            }
          }
          hero(episode: NEWHOPE) {
            name
          }
          r2d2 {
            name
          }
        }
      `;

      const result = await execute({
        schema: starwarsSchema,
        document: query,
      });

      expect(result).toMatchSnapshot();
    });
  });
});

describe('mocked', () => {
  const mockedSchema = builder.toSchema([Character, Droid, Episode, Human, MoreEpisodes, Query], {
    fieldDefinitions: [extraQueryFields],
    mocks: {
      Character: {
        name: () => 'C-3PO',
      },
    },
  });

  test('query some stuff', async () => {
    const query = gql`
      query {
        r2d2 {
          name
        }
      }
    `;

    const result = await execute({
      schema: mockedSchema,
      document: query,
    });

    expect(result).toMatchSnapshot();
  });
});
