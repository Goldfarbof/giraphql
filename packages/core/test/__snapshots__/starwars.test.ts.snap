// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mocked query some stuff 1`] = `
Object {
  "data": Object {
    "r2d2": Object {
      "name": "C-3PO",
    },
  },
}
`;

exports[`starwars example generates expected schema 1`] = `
"interface Character {
  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character!]!

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn(id: ID!): [Episode!]!
}

\\"\\"\\"A mechanical creature in the Star Wars universe.\\"\\"\\"
type Droid implements Character {
  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character!]!

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn(id: ID!): [Episode!]!

  \\"\\"\\"The primary function of the droid.\\"\\"\\"
  primaryFunction: String!
}

\\"\\"\\"One of the films in the Star Wars Trilogy\\"\\"\\"
enum Episode {
  \\"\\"\\"Released in 1977.\\"\\"\\"
  NEWHOPE

  \\"\\"\\"Released in 1980.\\"\\"\\"
  EMPIRE

  \\"\\"\\"Released in 1983\\"\\"\\"
  JEDI
}

\\"\\"\\"A humanoid creature in the Star Wars universe.\\"\\"\\"
type Human implements Character {
  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character!]!

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn(id: ID!): [Episode!]!

  \\"\\"\\"The home planet of the human, or null if unknown.\\"\\"\\"
  homePlanet: String
}

enum MoreEpisode {
  \\"\\"\\"Released in 1977.\\"\\"\\"
  NEWHOPE

  \\"\\"\\"Released in 1980.\\"\\"\\"
  EMPIRE

  \\"\\"\\"Released in 1983\\"\\"\\"
  JEDI
  OTHER
}

type Query {
  hero(
    \\"\\"\\"
    If omitted, returns the hero of the whole saga. If provided, returns the hero of that particular episode.
    \\"\\"\\"
    episode: Episode!
  ): Character!
  human(
    \\"\\"\\"id of the character\\"\\"\\"
    id: ID!
  ): Human!
  droid(
    \\"\\"\\"id of the character\\"\\"\\"
    id: ID!
  ): Droid!
  r2d2: Droid!
}
"
`;

exports[`starwars example queries query some stuff 1`] = `
Object {
  "data": Object {
    "droid": Object {
      "friends": Array [
        Object {
          "appearsIn": Array [
            "NEWHOPE",
            "EMPIRE",
            "JEDI",
          ],
          "homePlanet": "Tatooine",
          "name": "Luke Skywalker",
        },
        Object {
          "appearsIn": Array [
            "NEWHOPE",
            "EMPIRE",
            "JEDI",
          ],
          "homePlanet": null,
          "name": "Han Solo",
        },
        Object {
          "appearsIn": Array [
            "NEWHOPE",
            "EMPIRE",
            "JEDI",
          ],
          "homePlanet": "Alderaan",
          "name": "Leia Organa",
        },
      ],
      "name": "R2-D2",
      "primaryFunction": "Astromech",
    },
    "hero": Object {
      "name": "R2-D2",
    },
    "human": Object {
      "friends": Array [
        Object {
          "appearsIn": Array [
            "NEWHOPE",
          ],
          "name": "Wilhuff Tarkin",
        },
      ],
      "homePlanet": "Tatooine",
      "name": "Darth Vader",
    },
    "r2d2": Object {
      "name": "R2-D2",
    },
  },
}
`;
